common/authorization/rsaTokenKeyProvider.go

```diff
@@ -0,0 +1,150 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package authorization
+
+import (
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/loggerimpl"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/service/config"
+)
+
+type jwks struct {
+	Keys []jsonWebKeys `json:"keys"`
+}
+
+type jsonWebKeys struct {
+	Kty string   `json:"kty"`
+	Kid string   `json:"kid"`
+	Use string   `json:"use"`
+	N   string   `json:"n"`
+	E   string   `json:"e"`
+	X5c []string `json:"x5c"`
+}
+
+// Default RSA token key provider
+type rsaKeyProvider struct {
+	config config.JWTKeyProvider
+	keys   map[string]*rsa.PublicKey
+	timer  *time.Timer
+	logger log.Logger
+}
+
+var _ TokenKeyProvider = (*rsaKeyProvider)(nil)
+
+func NewRSAKeyProvider(cfg *config.Config) *rsaKeyProvider {
+	logger := loggerimpl.NewLogger(cfg.Log.NewZapLogger())
+	provider := rsaKeyProvider{config: cfg.Global.Security.JWTKeyProvider, logger: logger}
+	provider.init()
+	return &provider
+}
+
+func (a *rsaKeyProvider) init() {
+	a.keys = make(map[string]*rsa.PublicKey)
+	if len(a.config.KeySourceURIs) == 0 {
+		return // nothing to do
+	}
+	err := a.updateKeys()
+	if err != nil {
+		a.logger.Error("error during initial retrieval of RSA token keys: ", tag.Error(err))
+	}
+	a.timer = time.AfterFunc(a.config.RefreshTime, func() {
+		a.timerCallback()
+	})
+}
+
+func (a *rsaKeyProvider) rsaKey(alg string, kid string) (*rsa.PublicKey, error) {
+	if strings.ToLower(alg) != "rs256" {
+		return nil, fmt.Errorf("unexpected signing algorithm: %s", alg)
+	}
+
+	key, found := a.keys[kid]
+	if !found {
+		return nil, fmt.Errorf("RSA key not found for key ID: %s", kid)
+	}
+	return key, nil
+}
+
+func (a *rsaKeyProvider) timerCallback() {
+	err := a.updateKeys()
+	if err != nil {
+		a.logger.Error("error while refreshing RSA token keys: ", tag.Error(err))
+	}
+	a.timer = time.AfterFunc(a.config.RefreshTime, func() {
+		a.timerCallback()
+	})
+}
+
+func (a *rsaKeyProvider) updateKeys() error {
+	if len(a.config.KeySourceURIs) == 0 {
+		return fmt.Errorf("no URIs configured for retrieving keys")
+	}
+
+	keys := make(map[string]*rsa.PublicKey)
+
+	for _, uri := range a.config.KeySourceURIs {
+		resp, err := http.Get(uri)
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+
+		jwks := jwks{}
+		err = json.NewDecoder(resp.Body).Decode(&jwks)
+		if err != nil {
+			return err
+		}
+
+		for _, k := range jwks.Keys {
+			cert := "-----BEGIN CERTIFICATE-----\n" + k.X5c[0] + "\n-----END CERTIFICATE-----"
+			key, err := jwt.ParseRSAPublicKeyFromPEM([]byte(cert))
+			if err != nil {
+				return err
+			}
+			keys[k.Kid] = key
+		}
+	}
+	// swap old keys with the new ones
+	a.keys = keys
```

Comment:
Even though we are swapping map references, this is still technically not thread-safe and during test runs could cause random flaky failures because Go will "detect a data race." You'll want to use a Mutex here (maybe readonly vs write) to protect access to this variable as it is shared across go-routines

Reply:
Added a lock for it. I'm in disbelief that pointer swap is not considered atomic in Go.

Reply:
I think in this particular case, it's less about the pointer swap being atomic itself vs. letting go know there is a memory barrier. Similar to the volatile keyword in C# (https://stackoverflow.com/questions/8709918/why-this-program-does-not-go-into-infinite-loop-in-absence-of-volatility-of-a-bo), using some sort of synchronization primitive avoids the case where one goroutine makes a change to the pointer while another thread continues to read a stale/older copy of that pointer.

In practice, maybe it never happens, but who knows based on future runtime/compiler optimizations, etc

Reply:
This can be easily test with `-race` switch. `atomic.Value` fits better here. Example from the doc does literally the same thing: https://golang.org/pkg/sync/atomic/#example_Value_config